tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c

```diff
@@ -0,0 +1,457 @@
+		"Main test thread does not start in privilege mode\n");
+
+	/* Assert that the mode status variable indicates privilege mode */
+	zassert_true((_current->arch.mode & CONTROL_nPRIV_Msk) == 0,
+		"Thread nPRIV flag not clear for supervisor thread: 0x%0x\n",
+		_current->arch.mode);
+#endif /* CONFIG_USERSPACE */
+
+#if defined(CONFIG_FLOAT) && defined(CONFIG_FP_SHARING)
+	/* The main test thread is not (yet) actively using the FP registers */
+	zassert_true((_current->arch.mode & CONTROL_FPCA_Msk) == 0,
+		"Thread FPCA flag not clear at initialization 0x%0x\n",
+		_current->arch.mode);
+
+	/* Clear the thread's floating-point callee-saved registers' container.
+	 * The container will, later, be populated by the swap mechanism.
+	 */
+	memcpy(&_current->arch.preempt_float, 0,
+		sizeof(struct _preempt_float));
+
+	/* Randomize the FP callee-saved registers at test initialization */
+	load_fp_callee_saved_regs(&ztest_thread_fp_callee_saved_regs);
+
+	/* The main test thread is using the FP registers, but the .mode
+	 * flag is not updated until the next context switch.
+	 */
+	zassert_true((_current->arch.mode & CONTROL_FPCA_Msk) == 0,
+		"Thread FPCA flag not clear at initialization\n");
+#endif /* CONFIG_FLOAT && CONFIG_FP_SHARING */
+
+	/* Create an alternative (supervisor) testing thread */
+	k_thread_create(&alt_thread,
+		alt_thread_stack,
+		K_THREAD_STACK_SIZEOF(alt_thread_stack),
+		(k_thread_entry_t)alt_thread_entry,
+		NULL, NULL, NULL,
+		K_PRIO_COOP(PRIORITY), 0,
+		K_NO_WAIT);
+
+	/* Verify context-switch has not occurred. */
+	zassert_true(switch_flag == 0,
+		"Switch flag incremented when it should not have\n");
+
+	/* Prepare to force a context switch to the alternative thread,
+	 * by manually adding the current thread to the end of the queue,
+	 * so it will be context switched-out.
+	 */
+	int key;
+
+	key = irq_lock();
+	z_move_thread_to_end_of_prio_q(_current);
+	irq_unlock(key);
+
+	/* Verify context-switch has not occurred yet. */
+	zassert_true(switch_flag == 0,
+		"Switch flag incremented by unexpected context-switch.\n");
+
+	/* Store the callee-saved registers to some global memory
+	 * accessible to the alternative testing thread. That
+	 * thread is going to verify that the callee-saved regs
+	 * are successfully loaded into the thread's callee-saved
+	 * registers' container.
+	 */
+	__asm__ volatile (
+		"stmia %0, {r4-r11};\n\t"
+		:
+		: "r" (&ztest_thread_callee_saved_regs_container)
+		: "memory"
+	);
+
+	/* Manually trigger a context-switch to swap-out the current thread.
+	 * Request a return to a different interrupt lock state.
+	 */
+	__DMB();
+
+#if defined(CONFIG_NO_OPTIMIZATIONS)
+	SCB->ICSR |= SCB_ICSR_PENDSVSET_Msk;
+	__DSB();
+	__ISB();
+
+	/* The thread is now swapped-back in. */
+
+#else/* CONFIG_NO_OPTIMIZATIONS */
+
+	/* Fake a different irq_unlock key when performing swap.
+	 * This will be verified by the alternative test thread.
+	 */
+	register int swap_return_val __asm__("r0") =
```

Comment:
Why you are putting return value of __swap() in r0?

Reply:
so, I can test it later , and make sure the value is preserved